### ISSUE TITLE
Wheel support for python 3.9

### ISSUE THREAD

**Issue by jvesely** [numba/llvmlite#669]
Hi,

similar to other wheel requests, having a python 3.9 wheel would speed up and simplify (no need to pull llvm-devel or equivalent) python 3.9 setup.

thank you.

**Comment by seibert** [comment]
This is pending Python 3.9 support for Numba, which is more involved due to bytecode changes: https://github.com/numba/numba/issues/6345

**Comment by jvesely** [comment]
> This is pending Python 3.9 support for Numba, which is more involved due to bytecode changes: [numba/numba#6345](https://github.com/numba/numba/issues/6345)

Thanks! The comment thread indicates that llvmlite passes tests on Python 3.9.
Is there a chance llvmlite wheels will be released sooner than numba for projects that only use the former or is that too much maintenance trouble?

**Comment by esc** [comment]
@jvesely thank you for asking about this, but due to resource constraints this is unlikely.

**Comment by JackWetherell** [comment]
This issue is causing the sparse library (and I'm sure other packages with this as a dependency) to no longer install with python 3.9:

**Comment by esc** [comment]
@JackWetherell thank you for commenting on this. We are expecting to publish a release candidate for the next llvmlite version in 2-3 weeks. Hopefully this will include releasing llvmlite wheels for Python 3.9 on PyPi. Thank you for your patience.

**Comment by JackWetherell** [comment]
@esc Do you know if there is an expected time-frame for the llvmlie wheels being availible on pypy? Many thanks.

**Comment by esc** [comment]
@JackWetherell yes, release candidate wheels for llvmlite 0.36.0RC1 were released to PyPi last week, see our discourse for more information:

https://numba.discourse.group/t/numba-0-53-0-and-llvmlite-0-36-0-release-candidates/504/2

**Comment by esc** [comment]
And yes, there are wheels for Python 3.9 now too.

**Comment by JackWetherell** [comment]
Thank you for the information. It seems it still does not install for 3.9:
```
$ python -V
Python 3.9.1
```
```
$ pip -V      
pip 20.3.3 from /home/jack/.local/lib/python3.9/site-packages/pip (python 3.9)
```
```
$ pip install llvmile
Defaulting to user installation because normal site-packages is not writeable
ERROR: Could not find a version that satisfies the requirement llvmile
ERROR: No matching distribution found for llvmile
jack@holly:~ → pip install llvmlite
Defaulting to user installation because normal site-packages is not writeable
Collecting llvmlite
  Using cached llvmlite-0.35.0.tar.gz (121 kB)
Building wheels for collected packages: llvmlite
  Building wheel for llvmlite (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-mco757_y/llvmlite_6f7eefbe138b49ac85079e19081b04a9/setup.py'"'"'; __file__='"'"'/tmp/pip-install-mco757_y/llvmlite_6f7eefbe138b49ac85079e19081b04a9/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-a6fdvg1i
       cwd: /tmp/pip-install-mco757_y/llvmlite_6f7eefbe138b49ac85079e19081b04a9/
  Complete output (26 lines):
  running bdist_wheel
  /usr/bin/python /tmp/pip-install-mco757_y/llvmlite_6f7eefbe138b49ac85079e19081b04a9/ffi/build.py
  LLVM version... Traceback (most recent call last):
    File "/tmp/pip-install-mco757_y/llvmlite_6f7eefbe138b49ac85079e19081b04a9/ffi/build.py", line 105, in main_posix
      out = subprocess.check_output([llvm_config, '--version'])
    File "/usr/lib/python3.9/subprocess.py", line 420, in check_output
      return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
    File "/usr/lib/python3.9/subprocess.py", line 501, in run
      with Popen(*popenargs, **kwargs) as process:
    File "/usr/lib/python3.9/subprocess.py", line 947, in __init__
      self._execute_child(args, executable, preexec_fn, close_fds,
    File "/usr/lib/python3.9/subprocess.py", line 1819, in _execute_child
      raise child_exception_type(errno_num, err_msg, err_filename)
  FileNotFoundError: [Errno 2] No such file or directory: 'llvm-config'
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "/tmp/pip-install-mco757_y/llvmlite_6f7eefbe138b49ac85079e19081b04a9/ffi/build.py", line 191, in <module>
      main()
    File "/tmp/pip-install-mco757_y/llvmlite_6f7eefbe138b49ac85079e19081b04a9/ffi/build.py", line 181, in main
      main_posix('linux', '.so')
    File "/tmp/pip-install-mco757_y/llvmlite_6f7eefbe138b49ac85079e19081b04a9/ffi/build.py", line 107, in main_posix
      raise RuntimeError("%s failed executing, please point LLVM_CONFIG "
  RuntimeError: llvm-config failed executing, please point LLVM_CONFIG to the path for llvm-config
  error: command '/usr/bin/python' failed with exit code 1
  ----------------------------------------
  ERROR: Failed building wheel for llvmlite
  Running setup.py clean for llvmlite
Failed to build llvmlite
Installing collected packages: llvmlite
    Running setup.py install for llvmlite ... error
    ERROR: Command errored out with exit status 1:
     command: /usr/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-mco757_y/llvmlite_6f7eefbe138b49ac85079e19081b04a9/setup.py'"'"'; __file__='"'"'/tmp/pip-install-mco757_y/llvmlite_6f7eefbe138b49ac85079e19081b04a9/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-uu15m7v_/install-record.txt --single-version-externally-managed --user --prefix= --compile --install-headers /home/jack/.local/include/python3.9/llvmlite
         cwd: /tmp/pip-install-mco757_y/llvmlite_6f7eefbe138b49ac85079e19081b04a9/
    Complete output (29 lines):
    running install
    running build
    got version from file /tmp/pip-install-mco757_y/llvmlite_6f7eefbe138b49ac85079e19081b04a9/llvmlite/_version.py {'version': '0.35.0', 'full': 'ea23b026930cc00824c907172383f54c9d438e6b'}
    running build_ext
    /usr/bin/python /tmp/pip-install-mco757_y/llvmlite_6f7eefbe138b49ac85079e19081b04a9/ffi/build.py
    LLVM version... Traceback (most recent call last):
      File "/tmp/pip-install-mco757_y/llvmlite_6f7eefbe138b49ac85079e19081b04a9/ffi/build.py", line 105, in main_posix
        out = subprocess.check_output([llvm_config, '--version'])
      File "/usr/lib/python3.9/subprocess.py", line 420, in check_output
        return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
      File "/usr/lib/python3.9/subprocess.py", line 501, in run
        with Popen(*popenargs, **kwargs) as process:
      File "/usr/lib/python3.9/subprocess.py", line 947, in __init__
        self._execute_child(args, executable, preexec_fn, close_fds,
      File "/usr/lib/python3.9/subprocess.py", line 1819, in _execute_child
        raise child_exception_type(errno_num, err_msg, err_filename)
    FileNotFoundError: [Errno 2] No such file or directory: 'llvm-config'
    
    During handling of the above exception, another exception occurred:
    
    Traceback (most recent call last):
      File "/tmp/pip-install-mco757_y/llvmlite_6f7eefbe138b49ac85079e19081b04a9/ffi/build.py", line 191, in <module>
        main()
      File "/tmp/pip-install-mco757_y/llvmlite_6f7eefbe138b49ac85079e19081b04a9/ffi/build.py", line 181, in main
        main_posix('linux', '.so')
      File "/tmp/pip-install-mco757_y/llvmlite_6f7eefbe138b49ac85079e19081b04a9/ffi/build.py", line 107, in main_posix
        raise RuntimeError("%s failed executing, please point LLVM_CONFIG "
    RuntimeError: llvm-config failed executing, please point LLVM_CONFIG to the path for llvm-config
    error: command '/usr/bin/python' failed with exit code 1
    ----------------------------------------
ERROR: Command errored out with exit status 1: /usr/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-mco757_y/llvmlite_6f7eefbe138b49ac85079e19081b04a9/setup.py'"'"'; __file__='"'"'/tmp/pip-install-mco757_y/llvmlite_6f7eefbe138b49ac85079e19081b04a9/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-uu15m7v_/install-record.txt --single-version-externally-managed --user --prefix= --compile --install-headers /home/jack/.local/include/python3.9/llvmlite Check the logs for full command output.
```
Is this an unrelated issue?

**Comment by gmarkall** [comment]
> Thank you for the information. It seems it still does not install for 3.9: ...
> Is this an unrelated issue?

Does  the 0.36.0RC1 version install for you?

```
$ pip install llvmlite==0.36.0rc1
```

**Comment by esc** [comment]
@JackWetherell you have unfortunately hit the llvmlite infamous red-herring message, we are currently working on getting this resolved. The discussion is here: https://numba.discourse.group/t/fixing-filenotfounderror-errno-2-no-such-file-or-directory-llvm-config/510/2

In any case, I can help you debug this. Which platform are you on? I.e. operating system and cpu architecture? It may be that you need to be explicit about the exact version you install, as mentioned above by @gmarkall .

**Comment by JackWetherell** [comment]
@gmarkall @esc `$ pip install llvmlite==0.36.0rc1` works for me. But the problem is my dependency chain I am trying to resolve is sparse -> numba -> llvmlite. So I have now install llvmlite using the specific version, but numba still fails to install giving the llvmlite file not found message.

**Comment by esc** [comment]
@JackWetherell during the release candidate phase, you will need to install Numba with `pip install numba==0.53.0rc1.post1` for pip to pickup the correct version. For `pip install numba` to work OOTB on 3.9 you will have to wait for the final 0.53.0 release which should become available a few weeks from now, depending on how many bugs and issues are found with the release candidate.

**Comment by JackWetherell** [comment]
@esc Thank you very much for this. My question is regards to an open source project in development we hope to release with 3.9 support. Sparse is a key dependency of ours. I suppose I need to specify exactly `numba==0.53.0rc1.post1` as a dependency, and this will then support both 3.8 and 3.9 installations?

**Comment by esc** [comment]
@JackWetherell yes, that is correct in principle for the time being. However, I would advise against using `numba==0.53.0rc1.post1` as dependency specification inside a released package. This is a release candidate version meant primarily for testing and may very well still contains bugs and regressions (one has already been found). If you would like to release your project with 3.9 support and you rely on sparse/numba/llvmlite I would strongly advise waiting until the final release of Numba/llvmlite is published. If you absolutely can not wait, it may be OK to use the `numba==0.53.0rc1.post1` specification until a final release becomes available, but in that case, you should probably re-release your project with that version specification removed again. I.e. it may be OK as an interim solution. Hope that helps?

**Comment by esc** [comment]
@JackWetherell and to followup: `numba==0.53.0rc1.post1` will support Pythons 3.6,3.7,3.8 and 3.9.

**Comment by JackWetherell** [comment]
@esc Thank you again. There is no major time constraint so in this case we will wait until the final release is published.

**Comment by FedericoStra** [comment]
@JackWetherell I don't know if it is relevant to you, but I managed to get `llvmlite`/`numba` working on Fedora 33 WE with Python 3.9 by resolving the issue with `LLVM_CONFIG` not being set.

I installed LLVM 10, because the latest 11 is not supported by `llvmlite`:
```bash
$ sudo dnf install llvm10-devel
```

Then I installed `numba` (and indirectly `llvmlite` as a consequence) with the following command, which sets `LLVM_CONFIG` to the correct location:
```bash
$ sudo LLVM_CONFIG=`which llvm-config-10-64` pip install numba
Collecting numba
  Downloading numba-0.51.2.tar.gz (2.1 MB)
     |████████████████████████████████| 2.1 MB 1.2 MB/s 
Collecting llvmlite<0.35,>=0.34.0.dev0
  Downloading llvmlite-0.34.0.tar.gz (107 kB)
     |████████████████████████████████| 107 kB 1.2 MB/s 
Requirement already satisfied: numpy>=1.15 in /usr/local/lib64/python3.9/site-packages (from numba) (1.20.1)
Requirement already satisfied: setuptools in /usr/local/lib/python3.9/site-packages (from numba) (53.0.0)
Building wheels for collected packages: numba, llvmlite
  Building wheel for numba (setup.py) ... done
  Created wheel for numba: filename=numba-0.51.2-cp39-cp39-linux_x86_64.whl size=2850475 sha256=aca86c80283abf910bfdcea6998200eca839142a73a78ab4e46238ccd30e50c9
  Stored in directory: /root/.cache/pip/wheels/0a/51/41/9f5a61bcd9d25271d7021e05cbabcd574fb3d534d07654b780
  Building wheel for llvmlite (setup.py) ... done
  Created wheel for llvmlite: filename=llvmlite-0.34.0-cp39-cp39-linux_x86_64.whl size=1440134 sha256=dc522ef5b4731c9362c63b5fd8eee86bea1a1267e3ad04e753937894f56caaab
  Stored in directory: /root/.cache/pip/wheels/40/08/53/26580f3607587bd3fa1a18619841d1dcfedcabf2be52f8e2cd
Successfully built numba llvmlite
Installing collected packages: llvmlite, numba
Successfully installed llvmlite-0.34.0 numba-0.51.2
```

Everything seems to be working fine:
```python
$ python3.9
Python 3.9.1 (default, Jan 20 2021, 00:00:00) 
[GCC 10.2.1 20201125 (Red Hat 10.2.1-9)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import llvmlite, numba
>>> llvmlite.__version__
'0.34.0'
>>> numba.__version__
'0.51.2'
>>> @numba.jit
... def add(x, y): return x + y
... 
>>> add(3, 4) + add(3, 5.)
15.0
>>> add.signatures
[(int64, int64), (int64, float64)]

```

**Comment by JackWetherell** [comment]
@FedericoStra Thank you for the info. The `pip install numba==0.53.0rc1.post1` workaround seems to work for `sparse` on Arch Linux (defaults to 3.9). Thanks, Jack.

**Comment by esc** [comment]
@FedericoStra for what it is worth, we generally don't recommend to use LLVMs provided by linux distributions for use with llvmlite/Numba as we require various patches to be applied to LLVM for llvmlite/numba to work correctly in all circumstances. For more information, see: https://llvmlite.readthedocs.io/en/latest/faqs.html#why-static-linking-to-llvm

**Comment by esc** [comment]
We are also currently discussing the `llvm-config` not found error message and how it can be improved here: https://numba.discourse.group/t/fixing-filenotfounderror-errno-2-no-such-file-or-directory-llvm-config/510

**Comment by FedericoStra** [comment]
@esc Thank you for the info. I realize that my workaround is brittle and undesirable in a stable environment, but my goal was just being able to get back as soon as possible using numba and pycuda, hence for the moment I'm fine with a setup that will break anytime soon but works in the meantime. I see the intricacies of linking a custom build of LLVM and would like to help solve the current issue, but unfortunately this matter is too far out my field of expertise. Keep up the awesome work! ;)

**Comment by esc** [comment]
Closing this now. llvmlite wheels for Python 3.9 have become available with the 0.36.0 release.
